src/olympia/accounts/utils.py

```diff
@@ -71,3 +78,60 @@ def path_with_query(request):
 def camel_case(snake):
     parts = snake.split('_')
     return parts[0] + ''.join(part.capitalize() for part in parts[1:])
+
+
+def process_fxa_event(raw_body, **kwargs):
+    """Parse and process a single firefox account event."""
+    # Try very hard not to error out if there's junk in the queue.
+    log = getLogger('accounts.sqs')
+    event_type = None
+    try:
+        body = json.loads(raw_body)
+        event = json.loads(body['Message'])
+        event_type = event.get('event')
+        uid = event.get('uid')
+        timestamp = datetime.fromtimestamp(event.get('ts', ''))
+        if not (event_type and uid and timestamp):
+            raise ValueError(
+                'Properties event, uuid, and ts must all be non-empty')
+    except (ValueError, KeyError, TypeError), e:
+        log.exception('Invalid account message: %s' % e)
+    else:
+        if event_type == 'primaryEmailChanged':
+            email = event.get('email')
+            if not email:
+                log.error('Email property must be non-empty for "%s" event' %
+                          event_type)
+            else:
+                primary_email_change_event.delay(email, uid, timestamp)
+        else:
+            log.debug('Dropping unknown event type %r', event_type)
+
+
+def process_sqs_queue(queue_url, aws_region, queue_wait_time):
+    log = getLogger('accounts.sqs')
+    log.info('Processing account events from %s', queue_url)
+    try:
+        # Connect to the SQS queue.
+        sqs = boto3.client(
+            'sqs', aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
+            aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
+            region_name=aws_region)
+        # Poll for messages indefinitely.
+        while True:
+            # msg = queue.read(wait_time_seconds=queue_wait_time)
+            response = sqs.receive_message(
+                QueueUrl=queue_url, WaitTimeSeconds=queue_wait_time,
+                MaxNumberOfMessages=10)
+            msgs = response.get('Messages', []) if response else []
+            for message in msgs:
+                process_fxa_event(message.get('Body', ''))
```

Comment:
With `MaxNumberOfMessages` you can receive up to 10 messages or 10 change events. We should wrap the actual processing in a try/except block as well to avoid messages getting lost if one of the messages fail.

Also, I'd love to make `process_fxa_event` a celery task that can be retried in case of a failure (e.g database is down).

Reply:
> Also, I'd love to make process_fxa_event a celery task that can be retried in case of a failure (e.g database is down).

I wrote it like that at first but it was generating a large number of pointless celery tasks for all the unsupported `event_type`s (the queue gets everything FxA does, not just the relatively rare email change events).  So instead I moved just the database accessing part to a celery task - `primary_email_change_event`

Reply:
hmm, fair point, but let's at least wrap every `process_fxa_event` call so that we avoid any issues affecting other messages.

Reply:
https://github.com/mozilla/addons-server/pull/6812/commits/c608d981b2365ce3797b0852897e4a9d91db8706 - `process_fxa_event` was already pretty much `try`d inside the method but this will catch exceptions in the `delete_message` too.